### PR TITLE
Add lecture 4 material 

### PR DESCRIPTION
**Pull request by jeydobson**
Needs to go in after pull request #64. Still being worked on will remove draft status when ready.

### CODE REVIEW

**Comment by jeydobson** [comment]
Hi @UCLGuichard could you review this? Changes are for lecture 4 material. Needs one review so that I can then merge.

File: 03cpp2/sec02SmartPointers.md [hunk 1]
```diff
@@ -235,8 +212,14 @@ title: Smart Pointers
 * Always read the manual
 * Always consistently use it
 
+### Homework - 14
+
+* Create a memory leak using bare pointers and `new` to heap allocate and then repeat using using smart pointers
```
Comment (UCLGuichard) [inline 03cpp2/sec02SmartPointers.md]: A suggestion, could it be interesting to introduce tools like `valgrind` to consistently check for memory leaks ? Or at least suggest it ?

File: 03cpp2/sec02SmartPointers.md [hunk 2]
```diff
@@ -235,8 +212,14 @@ title: Smart Pointers
 * Always read the manual
 * Always consistently use it
 
+### Homework - 14
+
+* Create a memory leak using bare pointers and `new` to heap allocate and then repeat using using smart pointers 
+* Think of a use case where a resource should be uniqely owned and implement as a short application
```
Comment (UCLGuichard) [inline 03cpp2/sec02SmartPointers.md]: Uniquely

File: 03cpp2/sec03LambdaExpressions.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+
+* Create your own lambda expressions for each of the three basic syntax examples given above
+* Try to change a param from within, can you see a different behaviour if passed by reference or by value?
+* Use `std::count_if` with an appropriate lambda expression to count the number of values in a `vector<int>` that are divisable by 2 or 3   
+
+### Homework 16
+
+* Create a simple `Student` class that has public member variables storing `string firstname`, `string secondname` and `int age` 
+* Create a vector and fill it with varios instances of the `Student` class 
+* Create a sorting class `StudentSort` and that has a method `vector<Student> SortByAge(vector<Student> vs);` that returns a `vector<Student>` that has been sorted by age
```
Comment (UCLGuichard) [inline 03cpp2/sec03LambdaExpressions.md]: Various

File: 03cpp2/sec02SmartPointers.md (hunk 1 above)
Comment (jeydobson) [reply 03cpp2/sec02SmartPointers.md]: Good point, I've added a note that they will learn how to consistently check for memory leaks using `valgrind` in lecture 6

File: 03cpp2/sec02SmartPointers.md (hunk 2 above)
Comment (jeydobson) [reply 03cpp2/sec02SmartPointers.md]: Fixed in latest push

File: 03cpp2/sec03LambdaExpressions.md (hunk 3 above)
Comment (jeydobson) [reply 03cpp2/sec03LambdaExpressions.md]: Fixed in latest version

**Comment by jeydobson** [comment]
@UCLGuichard should be good to approve now

**Comment by UCLGuichard** [comment]
@jeydobson I will have a quick review and approve.

**Pull request merged**